### PR TITLE
Kinesis 8

### PR DESCRIPTION
**Pull request by fblundun**
No CHANGELOG yet as additional tickets may be added before release...

@alexanderdean could you publish rc1 please?

### CODE REVIEW

**Comment by alexanderdean** [comment]
CI/CD bot says:

> Build started!

**Comment by alexanderdean** [comment]
CI/CD bot says:

> Artifact published! https://bintray.com/artifact/download/snowplow/snowplow-generic/snowplow_kinesis_r80_southern_cassowary_rc1.zip

**Comment by alexanderdean** [comment]
@chuwy can you do code review on this? I know the Kinesis apps are new to you - feel free to grab @fblundun on a Hangout to go through together...

**Comment by chuwy** [comment]
@alexanderdean ok, I'll try.

File: 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/KinesisEnrichApp.scala [hunk 1]
```diff
@@ -250,7 +252,9 @@ object KinesisEnrichApp extends App {
       }
       case Some(other) => parser.usage(s"Enrichments argument [$other] must begin with 'file:' or 'dynamodb:'")
     }
-    """{"schema":"iglu:com.snowplowanalytics.snowplow/enrichments/jsonschema/1-0-0","data":[%s]}""".format(jsons.mkString(","))
+    val combinedJson = ("schema" -> "iglu:com.snowplowanalytics.snowplow/enrichments/jsonschema/1-0-0") ~
+    ("data" -> jsons.toList.map(parse(_)))
+    compact(render(combinedJson))
```
Comment (chuwy) [inline 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/KinesisEnrichApp.scala]: `render` is redundant here.

File: 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/KinesisEnrichApp.scala [hunk 2]
```diff
@@ -56,7 +56,9 @@ import scalaz.{Sink => _, _}
 import Scalaz._
 
 // json4s
+import org.json4s.JValue
```
Comment (chuwy) [inline 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/KinesisEnrichApp.scala]: Unused import. As well as `import java.net.URL`, `import java.util.Date`, `ConfigRenderOptions`, `import com.snowplowanalytics.snowplow.scalatracker.Tracker`.

**Comment by chuwy** [comment]
Shouldn't we update copyright years? At least in renamed `config.hocon.sample`?

File: 2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.snowplow.collectors/scalastream/ScalaCollectorApp.scala [hunk 3]
```diff
@@ -95,8 +103,23 @@ object ScalaCollector extends App {
...
   )
 
-  IO(Http) ! Http.Bind(handler,
-    interface=collectorConfig.interface, port=collectorConfig.port)
+  val bind = Http.Bind(
+    handler,
+    interface=collectorConfig.interface,
+    port=collectorConfig.port)
+
+  val bindResult = (IO(Http).ask(bind)(5.seconds)) flatMap {
```
Comment (chuwy) [inline 2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.snowplow.collectors/scalastream/ScalaCollectorApp.scala]: Redundant parenthesis  around `IO(Http).ask(bind)(5.seconds)`.

File: 2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.snowplow.collectors/scalastream/ScalaCollectorApp.scala [hunk 4]
```diff
@@ -95,8 +103,23 @@ object ScalaCollector extends App {
...
+    case b: Http.Bound => Future.successful(())
+    case failed: Http.CommandFailed => Future.failed(new RuntimeException(failed.toString))
+  }
+
+  bindResult onComplete {
+    case Success(s) =>
```
Comment (chuwy) [inline 2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.snowplow.collectors/scalastream/ScalaCollectorApp.scala]: Maybe `Suceess(_) => ()`?

**Comment by chuwy** [comment]
I also saw `return` keyword in `KinesisEnrichApp`. It actually quite safe there, but there's a very [dangerous misconception](https://tpolecat.github.io/2014/05/09/return.html) about it in Scala. I'd remove it everywhere I see it.

File: 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sinks/KinesisSink.scala [hunk 5]
```diff
@@ -311,6 +314,20 @@ class KinesisSink(provider: AWSCredentialsProvider,
...
+    badResponses.foldLeft(Map[String, (Long, String)]())((counts, r) => if (counts.contains(r.getErrorCode)) {
+      counts + (r.getErrorCode -> (counts(r.getErrorCode)._1 + 1 -> r.getErrorMessage))
+    } else {
+      counts + (r.getErrorCode -> (1, r.getErrorMessage))
+    })
+  }
+
+  private[sinks] def logErrorsSummary(errorsSummary: Map[String, (Long, String)]): Unit = {
+    for ((errorCode, (count, sampleMessage)) <- errorsSummary) {
+      error(s"$count records failed with error code ${errorCode}. Example error message: $sampleMessage")
```
Comment (chuwy) [inline 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sinks/KinesisSink.scala]: Redundant braces around `errorCode`. Or missing around `sampleMessage`?

File: 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sinks/KinesisSink.scala [hunk 6]
```diff
@@ -274,12 +277,12 @@ class KinesisSink(provider: AWSCredentialsProvider,
           val failurePairs = unsentRecords zip results filter { _._2.getErrorMessage != null }
           info(s"Successfully wrote ${unsentRecords.size-failurePairs.size} out of ${unsentRecords.size} records")
           if (failurePairs.size > 0) {
```
Comment (chuwy) [inline 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sinks/KinesisSink.scala]: `failurePairs.nonEmpty` would look nicer?

**Comment by chuwy** [comment]
@alexanderdean @fblundun LGTM! Nothing except style remarks.

**Comment by chuwy** [comment]
Also, are we going to use infinitive verb form for commit messages?

**Comment by fblundun** [comment]
Thanks @chuwy ! I agree with all your points and will make the necessary changes.

I have been doing imperative commit messages for a while, since it's recommended [here](http://stackoverflow.com/questions/3580013/should-i-use-past-or-present-tense-in-git-commit-messages). I think we are all supposed to be moving to using them @alexanderdean ?

**Comment by chuwy** [comment]
Good to hear. I also saw this [article](http://chris.beams.io/posts/git-commit/) and think this could very good decision. I also remember we talked about it few times.

**Comment by alexanderdean** [comment]
Yep we are moving to imperative, so the ticket name is identical to the commit message...

File: 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sinks/KinesisSink.scala [hunk 7]
```diff
@@ -274,12 +275,12 @@ class KinesisSink(provider: AWSCredentialsProvider,
...
-            failurePairs foreach { f => error(s"Record failed with error code [${f._2.getErrorCode}] and message [${f._2.getErrorMessage}]") }
+            val (unsentRecords, failedResults) = failurePairs.unzip
+            logErrorsSummary(getErrorsSummary(failedResults))
             backoffTime = getNextBackoff(backoffTime)
             error(s"Retrying all failed records in $backoffTime milliseconds...")
 
             val err = s"Failed to send ${failurePairs.size} events"
-            unsentRecords = failurePairs.map(_._1)
```
Comment (fblundun) [inline 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sinks/KinesisSink.scala]: The dangers of mutable state: I forgot that I was meant to modify the unsentRecords var and instead created a new unsentRecords val shadowing it.

**Comment by fblundun** [comment]
@alexanderdean could you publish rc2 please?

**Comment by alexanderdean** [comment]
CI/CD bot says:

> Build started!

**Comment by alexanderdean** [comment]
CI/CD bot says:

> Build failed:

```
[info] + return an invisible pixel
[info] x return a cookie expiring at the correct time
[error]    -10000 is not close to 0 +/- 2000 (CollectorServiceSpec.scala:143)
[info]
[info] + return the same cookie as passed in
[info] + return a P3P header
[info] + store the expected event as a serialized Thrift object in the enabled sink
[info] + report itself as healthy
[info]
[info]
[info] Total for specification CollectorServiceSpec
[info] Finished in 25 seconds, 877 ms
[info] 6 examples, 1 failure, 0 error
[info] PostSpec
[info]
[info] Snowplow's Scala collector should
[info] x return a cookie expiring at the correct time
[error]    -3000 is not close to 0 +/- 2000 (PostSpec.scala:138)
[info]
[info] + return the same cookie as passed in
[info] + return a P3P header
[info] x store the expected event as a serialized Thrift object in the enabled sink
[error]    1461314749929 is not close to 1461314755000 +/- 1000 (PostSpec.scala:179)
[info]
[info]
[info]
[error] Failed: Total 18, Failed 3, Errors 0, Passed 15
[error] Failed tests:
[error] 	com.snowplowanalytics.snowplow.collectors.scalastream.PostSpec
[error] 	com.snowplowanalytics.snowplow.collectors.scalastream.CollectorServiceSpec
```

**Comment by fblundun** [comment]
@alexanderdean this should be fixed by #2614 - could you try again please?

**Comment by alexanderdean** [comment]
```
[info] Snowplow's Scala collector should
[info] + return a cookie expiring at the correct time
[info] + return the same cookie as passed in
[info] + return a P3P header
[info] x store the expected event as a serialized Thrift object in the enabled sink
[error]    1461322125408 is not close to 1461322132000 +/- 1000 (PostSpec.scala:179)
[info]
[info]
[info]
[info] Total for specification PostSpec
[info] Finished in 30 seconds, 730 ms
[info] 4 examples, 1 failure, 0 error
[error] Failed: Total 18, Failed 1, Errors 0, Passed 17
[error] Failed tests:
[error] 	com.snowplowanalytics.snowplow.collectors.scalastream.PostSpec
```

**Comment by fblundun** [comment]
@alexanderdean could you try again please?

Also, is the test suite running particularly slowly for you? Based on your last comment, it looks like the PostSpec test took over 6 seconds to run.

**Comment by alexanderdean** [comment]
Trying again! The build didn't seem super-slow to me, maybe I missed that...

**Comment by alexanderdean** [comment]
The rc2 deployment will be handled by CI...

**Comment by alexanderdean** [comment]
Great work! Released

**Pull request merged**